batch/batch/front_end/front_end.py

```diff
@@ -855,10 +860,11 @@
             billing_project = record['billing_project']
             user = record['user']
             billing_projects[billing_project].append(user)
-    page_context = {
-        'billing_projects': billing_projects
-    }
-    return await render_template('batch', request, userdata, 'billing_projects.html', page_context)
+        page_context = {
+            'billing_projects': billing_projects
+        }
+        return await render_template('batch', request, userdata, 'billing_projects.html', page_context)
```

Comment:
So abstractly I like this, but we can't put things in the transaction for scoping reasons that we wouldn't want to otherwise.  So here I think the transaction needs to return billing data and the rendering happen outside.

This applies to a number of cases below.

Reply:
What's the issue with putting things in the transaction? None of that code will throw a sql exception.

Reply:
re: selects, I'll change them, but I will use FOR SHARE:
> FOR SHARE and LOCK IN SHARE MODE set shared locks that permit other transactions to read the examined rows but not to update or delete them. FOR SHARE and LOCK IN SHARE MODE are equivalent. However, FOR SHARE, like FOR UPDATE, supports NOWAIT, SKIP LOCKED, and OF tbl_name options. FOR SHARE is a replacement for LOCK IN SHARE MODE, but LOCK IN SHARE MODE remains available for backward compatibility.

Reply:
OK, I added all the ones that seem necessary

Reply:
I think 5.7 doesn't have SELECT FOR SHARE.

Reply:
> What's the issue with putting things in the transaction? None of that code will throw a sql exception.

It will hold the lock longer than necessary decreasing the concurrency available in the database.  You should always make the critical sections as short as possible.  I find this extra code in the critical section literally painful.  I will make a PR to fix these.